content/blog/2020-12-19-asmm/index.md

```diff
@@ -0,0 +1,119 @@
++++
+title = "Automatic Static Memory Management in Bril"
+[extra]
+bio = """
+  Patrick LaFontaine is a senior undergraduate student interested in programming language design and implementation. He also enjoys backpacking and attempting to bake the perfect loaf of bread.
+"""
+[[extra.authors]]
+name = "Patrick LaFontaine"
++++
+
+## Motivation: Compiler-led Garbage Collection
+
+The [memory extension][memex] for the Bril programming language provides manually-managed memory for the allocation of arrays. Manual memory management, when the programmer is in charge of explicitly freeing their memory, provides full control to the programmer at the cost of exposing them to difficult-to-debug memory bugs. The common bugs are leaking memory by forgetting to call ```free```, attempting to ```free``` memory more than once, or attempting to use a memory location after it has been freed. A popular way to handle these issues is to have a run-time system in charge of deciding when to ```free``` memory. These run-time garbage collectors allow programmers to easily work with memory, but often at the cost of worse performance, poor latency with "stop the world" pauses or significant memory usage. One alternative to this is to put the compiler in charge of deciding when to free memory. This handles all of the memory management logic, at compile time, for the programmer, hence "Automatic Static Memory Management". The main downside of this approach is that this static analysis, like all compiler analyses, must be conservative which can limit the expressiveness of the programming language.
```

Comment:
Normally, it's "runtime system" rather than "run-time system." But certainly "run-time garbage collector." https://capra.cs.cornell.edu/styleguide/#runtime